### DISCUSSION TITLE
Autenite reconstruction using MTEX 5.7.0 and 5.9.0

### DISCUSSION THREAD

**Discussion by ArianeNeves** [mtex-toolbox/mtex#1911] [Ask Anything]
Hi! @nyyssont, @frankNiessen and @AzdiarGazder,

Please, I reconstructed the austenite in biphasic steel using Mtex 5.7.0 and 5.9.0. First, I did the partition of ferrite phase and use the martensite data to do the reconstruction. The results are showed below.

![Apresentação1](https://github.com/mtex-toolbox/mtex/assets/81368454/2c6871a2-6d8b-4920-a1cd-6f1502988c12)

Using Mtex 5.9.0, we observe child grains with lower probability. How can I improve the result using the MTEX 5.9.0? May you can help me?

Ariane

**Comment by frankNiessen** [comment]
Hi Ariane, 

thank you for your inquiry. Do you work with exactly the same script and the same data in MTEX 5.7 and MTEX 5.9 and observe different results? 

The probability is indeed quite low for some of your martensitic clusters which is almost always caused by a poor representation of the orientation relationship between austenite and martensite. This implies that your orientation relationship used for parent grain reconstruction does not match the local grain boundary misorientations in your map which is likely caused by one of the following reasons:

1. Your microstructure does not follow a consistent orientation relationship (maybe your data does not only consider martensite - or different martensite morphologies?)
2. The orientation relationship was not optimized for the particular microstructure (you might be running with a rational orientation relationship like Kurdjumov-Sachs?)

If you can provide us with some more information we should be able to get this solved.

Best wishes
Frank

**Comment by AzdiarGazder** [comment]
Hi @ArianeNeves,

As Frank mentioned, we need more information (for example, your EBSD map data and your script) before we resolve this issue.

In the meantime, if you'd like to have an initial go at issue resolution, I'd suggest the following:

1. Please use the latest stable releases of **[ORTools v2.2.0](https://github.com/ORTools4MTEX/ORTools/releases/tag/v2.2.0)** and **[MTEX v5.10](https://mtex-toolbox.github.io/download.html)** together.
OR
 Alternatively, you could use the latest developer releases of **[ORTools](https://github.com/ORTools4MTEX/ORTools)** and **[MTEX](https://github.com/mtex-toolbox/mtex/tree/develop)** together.

2. Please apply the comment and script uploaded at:
https://github.com/ORTools4MTEX/ORTools/discussions/46#discussioncomment-6410242
It finds the optimal OR via peak-fitting and as Frank mentioned in his point 2, may be just the thing that needs doing to resolve your issue.

Warm regards,
Azdi

**Comment by nyyssont** [comment]
I do not have much to add to the above answers. There is no significant difference between the `calcGBVotes` functions of 5.7.0 and 5.9.0 when it comes to calculating parent orientations from child orientations. 

As both Frank and Azdi point out, a good start towards determining the issue would be to examine if there is any difference in the optimized orientation relationship generated by 5.7.0 and 5.9.0.

Another thing - your child grains seem to be rather large on occasion, when compared to the pixel-level orientations. What threshold are you using to construct the child grain map?

-Tuomo

**Comment by ArianeNeves** [comment]
Thanks for the answer, @frankNiessen, @AzdiarGazder and @nyyssont!

The .ang file and the codes are attached. I used diferrents code for the reconstruction using the version 5.7.0 and 5.9.0. The quantile (calcParent2Child) was 0.75 for both versions. 

[Sample1ang.txt](https://github.com/mtex-toolbox/mtex/files/12233215/Sample1ang.txt)
[ReconstructionMTEX5.7.0.txt](https://github.com/mtex-toolbox/mtex/files/12233229/ReconstructionMTEX5.7.0.txt)
[ReconstructionMTEX5.9.0.txt](https://github.com/mtex-toolbox/mtex/files/12233231/ReconstructionMTEX5.9.0.txt)

I observed this when I reconstructed the ferrite and martensite microstructure. I did the ferrite partition before the reconstruction. Then the ferrite regions are non-indexed pixels for MTEX. In addition, the packets, Bain groups and block width are larger for this material. 

The threshold used to construct the child grain map was 3°. I used the function job.calcParent2Child to optimize the orientation relationship. I did not find difference between the orientation relationship used to reconstruct in 5.7.0 and 5.9.0 versions. 

Now I will check the point 1 cited by @AzdiarGazder.

Thanks for the help!

Ariane

**Comment by AzdiarGazder** [comment]
Hi @ArianeNeves,

I ran the script I mentioned in my earlier post using your *.ang file. 

1. I believe the map may be of a duplex steel with the uninidexed regions being austenite.
So I'm a bit confused as to why the austenite was removed from the map in the first place as there is no need for that step at all.
In any case, please try and run the script with all phases in place.
With austenite in the map, you could then use the use the peak fitter to find the initial OR.

2. Following on from the above point, since there is no parent phase in the map you uploaded, the peak fitter function cannot be used to find the initial OR.
The only option is to provide an initial best guess OR to begin the reconstruction process. 
Here, I found the best initial guess is the inverse Greninger-Troiano OR.

Hope this helps.

Warm regards,
Azdi

**Comment by ArianeNeves** [comment]
Hi @AzdiarGazder, 

The unindexed regions is ferrite phase (CCC). If I keep the ferrite in the dataset, Mtex reconstructs the ferrite as well, because ferrite and martensite have the same crystalline structure. This also influence the the optimized orientation relationship. 

Thanks for the help!

Ariane

**Comment by nyyssont** [comment]
Some thoughts on this particular case:

The reconstruction algorithms give probability values to pairs of neighboring grains, with the probability signifying the likelihood that a mutual parent orientation is shared by the grains. The default option is to use the average orientations of grains to make the necessary calculations. This is done to conserve memory. However, this choice becomes problematic when a "grain" as determined by MTEX is actually composed of several martensitic variants with mutual misorientation angles under 3 degrees. The misorientation calculated from average orientations of these clumps of variants are then a poor representation of the actual misorientations at the boundaries.

To counteract this, the option `job.useBoundaryOrientations` can be set to `true`. The misorientations used to determine probabilities are then calculated from the individual orientation pixels at the boundaries. In your case, I would recommend using this option.

Another thing to do in cases like this is to use a generous threshold value for the variant graph. I generally use 5 degrees.

A third thing to do is to decrease the quantile value. This concentrates the p2c optimization to "good" misorientations in a case where the situation is muddled by the variant-clump grains.

I tried the following script which seems to give a better result on MTEX 5.10.0:

```
p2c = orientation.KurdjumovSachs(csParent,csChild);
job = parentGrainReconstructor(ebsd,grains,p2c)
job.useBoundaryOrientations = true;

job.calcParent2Child('quantile',0.6)

% Variant Graph based parent grain reconstruction
job.calcVariantGraph('threshold',5*degree,'tolerance',2*degree)
job.clusterVariantGraph
job.calcParentFromVote('minProb',0.01)
```

As you see, I did not split the variant graph algorithm into two sections. This possibility exists for cases where the maps are very large and memory needs to be conserved. For maps of this size, splitting the variant graph operation is not necessary.

**Comment by AzdiarGazder** [comment]
Hi @ArianeNeves,

I am just following up to find out if your issue was resolved.
If so, please mark the answer that proved most useful in this thread.

I was also having a think about how you were presenting the EBSD map for reconstruction; where 1 phase/micro-constituent (ferrite) was manually deleted due to it being crystallographically the same as the one you wanted to reconstruct (martensite). 

Deleting ebsd map data may cause downstream issues with reconstruction (especially fragmentation of the child grains).
The easiest way to avoid such unnecessary errors from creeping in would be to start with a complete EBSD map (i.e. - without deleting or removing any phases).

The easiest way to do that is as follows:

Step 1: Segment the ferrite and martensite phases in your map.
Scripts showing you how to do this are shown [here](https://github.com/AzdiarGazder/mtexTools/tree/main/ferriteQuantifier) and [here](https://github.com/mtex-toolbox/mtex/issues/1929).

Step 2: Begin parent grain reconstruction of ONLY the martensite phase in your map.

Hope this helps.

Warm regards,
Azdi

**Comment by ArianeNeves** [comment]
Hi @AzdiarGazder!
I did not find a answer for this issue. 

I did the segmentation in the OIM software. I tried to do the reconstruction with the complete dataset. The otimized orientation relationship was different when we compare to martensite dataset. 

What intrigues me is that MTEX 5.7.0 reconstruct the all austenite grains in the map. 

What are you suggesting is segmenting the ferrite phase using MTEX?

Warm regards,

Ariane

**Reply by AzdiarGazder** [reply]
Hi @ArianeNeves,

What I am suggesting is that you segment the bcc pixels in your map into ferrite and martensite phases before you begin the parent grain reconstruction of martensite.
It does not matter if the segmentation is done in OIM or MTEX - just so long as it is done before you begin reconstruction.
In doing so, you present a complete map (i.e.- one without large blank areas) for the reconstruction step.
The segmentation into different phases helps MTEX know which pixels to reconstruct and which to leave untouched.

Once you've done the phase segmentation:
1. Please use the latest stable releases of **[ORTools v2.2.0](https://github.com/ORTools4MTEX/ORTools/releases/tag/v2.2.0)** and **[MTEX v5.10](https://mtex-toolbox.github.io/download.html)** together.
OR
 Alternatively, you could use the latest developer releases of **[ORTools](https://github.com/ORTools4MTEX/ORTools)** and **[MTEX](https://github.com/mtex-toolbox/mtex/tree/develop)** together.

2. Please apply the comment and script uploaded at:
https://github.com/ORTools4MTEX/ORTools/discussions/46#discussioncomment-6410242
It finds the optimal OR via peak-fitting and may be just the thing that needs doing to resolve your issue.

Warm regards,
Azdi

**Comment by ArianeNeves** [comment]
Hi @AzdiarGazder!
It was exactly what I did.  The ferrite was segmented before in the OIM. So, the ferrite points were non-indexed points in the dataset used in the MTEX. In the Figure showed below, the white area is ferrite after the reconstruction using MTEX 5.7.0. On the other hand, the black area is ferrite and the white area is non-reconstructed austenite when I used MTEX 5.9.0.

![256823912-2c6871a2-6d8b-4920-a1cd-6f1502988c12](https://github.com/mtex-toolbox/mtex/assets/81368454/5c229794-c6b1-4c5c-844b-96f159218a2f)

Best regards,
Ariane

**Reply by AzdiarGazder** [reply]
Hi @ArianeNeves,

I understand you have segmented the map in OIM into ferrite and martensite.
But you are manually **deleting** or **removing** the pixels corresponding to the ferrite phase before starting reconstruction in MTEX.
So you are presenting a map full of non-indexed regions to MTEX for reconstruction.
I am requesting you not to do that.

Instead, as the last step of segmentation in OIM, I suggest you **reassign & rename** the pixels belonging to the **ONE** bcc phase as **TWO separate & distinct** bcc phases called *ferrite* and *martensite*.
In this way, the *.ang file of your map **before** reconstruction has 4 "phases" in it instead of 3:
1. non-indexed 
2. Iron alpha (which you can rename as ferrite)
3. Iron gamma (which you can rename as austenite)
4. Martensite (whose CS is also iron alpha but in this case, the phase name and number is different)

**Reassigning and renaming** the bcc phase as 2 separate and distinct phases means your map does not have large holes with non-indexed regions in it. You can then point MTEX to only reconstruct the martensite phase and leave the ferrite phase alone. This makes reconstruction much easier and prevents any unintentional errors from creeping into the reconstruction.

Lastly, please use the latest versions of the MTEX and ORTools software suites (preferably the developer versions).
That way you can benefit from the latest changes/improvements/updates.

Warm regards,
Azdi
